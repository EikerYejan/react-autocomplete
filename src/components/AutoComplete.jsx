import React, { createRef } from 'react'
import './AutoComplete.css'
import Option from './Option/Option'

/* Props
 - placeholder
 - common attributes
 - options
*/

const options = [
  { label: 'Option 1', value: 'OPT1' },
  { label: 'Option 2', value: 'OPT2' },
  { label: 'Option 3', value: 'OPT3' },
  { label: 'Option 4', value: 'OPT4' },
  { label: 'Option 5', value: 'OPT5' },
  { label: 'Option 6', value: 'OPT6' },
  { label: 'Option 7', value: 'OPT7' },
  { label: 'Option 8', value: 'OPT8' },
]

class AutoComplete extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      showOptions: props?.defaultOpen ?? false,
    }
    this.wrapperRef = createRef(null)
  }

  onWindowClick = (e) => {
    if (this.wrapperRef?.current && !this.wrapperRef.current.contains(e.target)) {
      return this.setState((prev) => ({ ...(prev ?? {}), showOptions: false }))
    }
  }

  onHandleClick = () => {
    return this.setState((prev) => ({ ...(prev ?? {}), showOptions: !prev.showOptions }))
  }

  onOptionClick = (selectedOption) => {
    return this.setState((prev) => ({ ...(prev ?? {}), searchTerm: selectedOption, showOptions: false }))
  }

  onSearchChange = (e) => {
    return this.setState((prev) => ({ ...(prev ?? {}), searchTerm: e?.target?.value, showOptions: true }))
  }

  componentDidMount() {
    window.addEventListener('mousedown', this.onWindowClick)
  }

  componentWillUnmount() {
    window.removeEventListener('mousedown', this.onWindowClick)
  }

  render() {
    const { showOptions, searchTerm } = this.state

    return (
      <div ref={this.wrapperRef} className="AutoComplete__">
        <p className="AutoComplete__label">Country</p>
        <div className="AutoComplete__handle">
          <input
            onChange={this.onSearchChange}
            value={searchTerm}
            placeholder="Type your country"
            type="text"
            className="AutoComplete__input"
          />
          <button onClick={this.onHandleClick} type="button" title="Expand" class="AutoComplete__expand">
            <img src={`/images/${showOptions ? 'expand_up' : 'expand'}.svg`} alt="expand" />
          </button>
        </div>
        {showOptions && (
          <ul className="AutoComplete__options">
            {options.map((option) => (
              <Option
                isSelected={option.label === searchTerm || option.value === searchTerm}
                onClick={() => this.onOptionClick(option.label)}
                key={option.value}
                {...option}
              />
            ))}
          </ul>
        )}
      </div>
    )
  }
}

export default AutoComplete
