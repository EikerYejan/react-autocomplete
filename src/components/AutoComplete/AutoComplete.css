@keyframes optionsMount {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

.AutoComplete__ {
  width: 350px;
  max-width: 350px;
  margin: 50px auto;
  color: var(--blue);
  position: relative;
}

.AutoComplete__loader {
  position: absolute;
  top: 11px;
  right: 12px;
}

.AutoComplete__label {
  font-weight: 600;
  font-size: 16px;
  margin-bottom: 8px;
}

.AutoComplete__handle {
  position: relative;
  width: 100%;
  height: 48px;
  background-color: var(--white);
}

.AutoComplete__input {
  width: 100%;
  height: 100%;
  border-radius: 4px;
  border: 1px solid var(--borderColor);
  display: flex;
  justify-content: flex-start;
  align-items: center;
  background-color: transparent;
  padding: 0;
  padding-left: 15px;
}

.AutoComplete__input:disabled {
  background-color: var(--lightGrey);
}

.AutoComplete__expand {
  position: absolute;
  right: 0;
  top: 15%;
  width: 40px;
  height: 70%;
  border: none;
  border-left: 1px solid var(--borderColor);
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 0;
  background: none;
  cursor: pointer;
}

.AutoComplete__expand img {
  width: 20px;
  height: 20px;
  display: block;
  margin: auto;
}

.AutoComplete__options {
  padding: 0;
  list-style: none;
  margin-top: 5px;
  border: 1px solid var(--borderColor);
  border-bottom: none;
  max-height: 200px;
  overflow-x: hidden;
  overflow-y: scroll;
  box-shadow: 0px 2px 12px rgba(0, 0, 0, 0.15);
  border-radius: 4px;
  animation: optionsMount 0.2s ease-in-out;
  position: absolute;
  width: 100%;
  z-index: 1;
  background-color: var(--white);
}

.AutoComplete__empty-options {
  height: 80px;
  display: flex;
  justify-content: center;
  align-items: center;
}
