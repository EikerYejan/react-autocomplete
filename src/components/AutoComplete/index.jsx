import React, { createRef } from 'react'
import Option from '../Option/Option'
import Loader from '../Loader'
import './AutoComplete.css'
import Skeleton from '../Skeleton/Skeleton'

class AutoComplete extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      showOptions: props?.defaultOpen ?? false,
    }
    this.wrapperRef = createRef(null)
    this.debounceTimeout = createRef()
  }

  onWindowClick = (e) => {
    if (this.wrapperRef?.current && !this.wrapperRef.current.contains(e.target)) {
      return this.setState((prev) => ({ ...(prev ?? {}), showOptions: false }))
    }
  }

  onHandleClick = () => {
    return this.setState((prev) => ({ ...(prev ?? {}), showOptions: !prev.showOptions }))
  }

  onOptionClick = (selectedOption) => {
    if (this.props?.onSelect) this.props.onSelect(selectedOption)

    return this.setState((prev) => ({ ...(prev ?? {}), searchTerm: selectedOption.label, showOptions: false }))
  }

  onSearchChange = (e) => {
    const { value } = e.target
    this.setState((prev) => ({
      ...prev,
      searchTerm: value && value?.length > 0 ? value : undefined,
    }))

    if (this.debounceTimeout.current) clearTimeout(this.debounceTimeout.current)
    this.debounceTimeout.current = setTimeout(this.onSearchFinish, 500, value)
  }

  onSearchFinish = async (value) => {
    const matchedOptions = await this.props.onSearch(value)

    return this.setState((prev) => ({
      ...prev,
      showOptions: true,
      matchedOptions,
    }))
  }

  componentDidMount() {
    window.addEventListener('mousedown', this.onWindowClick)
  }

  componentWillUnmount() {
    window.removeEventListener('mousedown', this.onWindowClick)
  }

  render() {
    const { showOptions, searchTerm, matchedOptions } = this.state
    const { label, placeholder, options, loading } = this.props

    const optionsToShow = typeof searchTerm === 'string' && typeof searchTerm.length > 0 ? matchedOptions : options

    const renderContent = () => {
      if (loading)
        return Array(9)
          .fill(0)
          .map((_, i) => <Skeleton key={i} />)

      if (!optionsToShow || optionsToShow.length === 0)
        return (
          <li data-testid="autocomplete-no-data-item" className="AutoComplete__empty-options">
            No data available
          </li>
        )

      return optionsToShow.map((option) => (
        <Option
          searchTerm={searchTerm}
          data-testid="autocomplete-option"
          isSelected={option.label === searchTerm || option.value === searchTerm}
          onClick={() => this.onOptionClick(option)}
          key={option.value}
          {...option}
        />
      ))
    }

    return (
      <div ref={this.wrapperRef} className="AutoComplete__">
        <p className="AutoComplete__label">{label}</p>
        <div className="AutoComplete__handle">
          <input
            onChange={this.onSearchChange}
            value={searchTerm}
            placeholder={placeholder}
            type="text"
            className="AutoComplete__input"
            data-testid="autocomplete-input"
          />
          {loading && <Loader data-testid="autocomplete-loader" className="AutoComplete__loader" />}
          <button
            data-testid="autocomplete-handle"
            onClick={this.onHandleClick}
            type="button"
            title="Expand"
            className="AutoComplete__expand"
          >
            <img src={`/images/${showOptions ? 'expand_up' : 'expand'}.svg`} alt="expand" />
          </button>
        </div>
        {showOptions && <ul className="AutoComplete__options">{renderContent()}</ul>}
      </div>
    )
  }
}

export default AutoComplete
