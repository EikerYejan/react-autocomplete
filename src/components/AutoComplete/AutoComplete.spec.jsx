import React from 'react'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import AutoComplete from '.'

const options = [
  { label: 'Option 1', value: 'OPT1' },
  { label: 'Option 4', value: 'OPT4' },
  { label: 'Option 5', value: 'OPT5' },
  { label: 'Option 6', value: 'OPT6' },
  { label: 'Option 7', value: 'OPT7' },
  { label: 'Option 8', value: 'OPT8' },
]
const onSearch = jest.fn(() => [])
const onSelect = jest.fn()
const placeHolder = 'Test'
const label = 'Label'

const props = { options, onSearch, onSelect, placeHolder, label }

jest.useFakeTimers()
jest.spyOn(global, 'setTimeout')

describe('<AutoComplete />', () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  it('Should render', () => {
    const { baseElement } = render(<AutoComplete {...props} />)

    expect(baseElement).toBeInTheDocument()
  })

  it('Should render loader', () => {
    const { getByTestId } = render(<AutoComplete {...props} loading />)
    const loader = getByTestId('autocomplete-loader')

    expect(loader).toBeInTheDocument()
  })

  it('Should type text', async () => {
    const { getByTestId } = render(<AutoComplete {...props} />)
    const input = getByTestId('autocomplete-input')

    userEvent.type(input, 'S')

    jest.runAllTimers()
    expect(onSearch).toHaveBeenCalledWith('S')
  })

  it('Should render "No Data" component', () => {
    const { options, ...componentProps } = props
    const { getByTestId } = render(<AutoComplete {...componentProps} options={[]} defaultOpen />)
    const element = getByTestId('autocomplete-no-data-item')

    expect(element).toBeInTheDocument()
  })

  it('Should render options list', () => {
    const { getAllByTestId } = render(<AutoComplete {...props} defaultOpen />)
    const options = getAllByTestId('autocomplete-option')
    const option = options[0]

    expect(options.length).toEqual(props.options.length)
    userEvent.click(option)

    expect(onSelect).toHaveBeenCalledWith(props.options[0])
  })

  it('Should show options list when clicking in the handle', () => {
    const { baseElement, getByTestId } = render(<AutoComplete {...props} />)
    const handle = getByTestId('autocomplete-handle')
    userEvent.click(handle)

    expect(baseElement).toMatchSnapshot()
  })
})
