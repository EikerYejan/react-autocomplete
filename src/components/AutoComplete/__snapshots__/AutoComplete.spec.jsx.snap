// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoComplete /> Should show options list when clicking in the handle 1`] = `
<body>
  <div>
    <div
      class="AutoComplete__"
    >
      <p
        class="AutoComplete__label"
      >
        Label
      </p>
      <div
        class="AutoComplete__handle"
      >
        <input
          class="AutoComplete__input"
          data-testid="autocomplete-input"
          type="text"
          value=""
        />
        <button
          class="AutoComplete__expand"
          data-testid="autocomplete-handle"
          title="Expand"
          type="button"
        >
          <img
            alt="expand"
            src="/images/expand_up.svg"
          />
        </button>
      </div>
      <ul
        class="AutoComplete__options"
      >
        <li
          class="option false"
          data-testid="autocomplete-option"
          title="Option 1"
          value="OPT1"
        >
          <span
            class="option-highlight"
          >
             
          </span>
          Option 1
        </li>
        <li
          class="option false"
          data-testid="autocomplete-option"
          title="Option 4"
          value="OPT4"
        >
          <span
            class="option-highlight"
          >
             
          </span>
          Option 4
        </li>
        <li
          class="option false"
          data-testid="autocomplete-option"
          title="Option 5"
          value="OPT5"
        >
          <span
            class="option-highlight"
          >
             
          </span>
          Option 5
        </li>
        <li
          class="option false"
          data-testid="autocomplete-option"
          title="Option 6"
          value="OPT6"
        >
          <span
            class="option-highlight"
          >
             
          </span>
          Option 6
        </li>
        <li
          class="option false"
          data-testid="autocomplete-option"
          title="Option 7"
          value="OPT7"
        >
          <span
            class="option-highlight"
          >
             
          </span>
          Option 7
        </li>
        <li
          class="option false"
          data-testid="autocomplete-option"
          title="Option 8"
          value="OPT8"
        >
          <span
            class="option-highlight"
          >
             
          </span>
          Option 8
        </li>
      </ul>
    </div>
  </div>
</body>
`;
